pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/MissingConsumingSegmentFinder.java

```diff
@@ -0,0 +1,221 @@
+    MissingSegmentInfo info = findMissingSegments(idealState.getRecord().getMapFields(), Instant.now());
+    _controllerMetrics.setValueOfTableGauge(_realtimeTableName, ControllerGauge.MISSING_CONSUMING_SEGMENT_TOTAL_COUNT,
+        info._totalCount);
+    _controllerMetrics
+        .setValueOfTableGauge(_realtimeTableName, ControllerGauge.MISSING_CONSUMING_SEGMENT_NEW_PARTITION_COUNT,
+            info._newPartitionGroupCount);
+    _controllerMetrics
+        .setValueOfTableGauge(_realtimeTableName, ControllerGauge.MISSING_CONSUMING_SEGMENT_MAX_DURATION_MINUTES,
+            info._maxDurationInMinutes);
+  }
+
+  @VisibleForTesting
+  MissingSegmentInfo findMissingSegments(Map<String, Map<String, String>> idealStateMap, Instant now) {
+    // create the maps
+    Map<Integer, LLCSegmentName> partitionGroupIdToLatestConsumingSegmentMap = new HashMap<>();
+    Map<Integer, LLCSegmentName> partitionGroupIdToLatestCompletedSegmentMap = new HashMap<>();
+    idealStateMap.forEach((segmentName, instanceToStatusMap) -> {
+      LLCSegmentName llcSegmentName = LLCSegmentName.of(segmentName);
+      if (llcSegmentName != null) { // Skip the uploaded realtime segments that don't conform to llc naming
+        if (instanceToStatusMap.containsValue(SegmentStateModel.CONSUMING)) {
+          updateMap(partitionGroupIdToLatestConsumingSegmentMap, llcSegmentName);
+        } else if (instanceToStatusMap.containsValue(SegmentStateModel.ONLINE)) {
+          updateMap(partitionGroupIdToLatestCompletedSegmentMap, llcSegmentName);
+        }
+      }
+    });
+
+    MissingSegmentInfo missingSegmentInfo = new MissingSegmentInfo();
+    if (!_partitionGroupIdToLargestStreamOffsetMap.isEmpty()) {
+      _partitionGroupIdToLargestStreamOffsetMap.forEach((partitionGroupId, largestStreamOffset) -> {
+        if (!partitionGroupIdToLatestConsumingSegmentMap.containsKey(partitionGroupId)) {
+          LLCSegmentName latestCompletedSegment = partitionGroupIdToLatestCompletedSegmentMap.get(partitionGroupId);
+          if (latestCompletedSegment == null) {
+            // There's no consuming or completed segment for this partition group. Possibilities:
+            //   1) it's a new partition group that has not yet been detected
+            //   2) the first consuming segment has been deleted from ideal state manually
+            missingSegmentInfo._newPartitionGroupCount++;
+            missingSegmentInfo._totalCount++;
+          } else {
+            // completed segment is available, but there's no consuming segment
+            SegmentZKMetadata segmentZKMetadata = _segmentMetadataFetcher
+                .fetchSegmentZkMetadata(_realtimeTableName, latestCompletedSegment.getSegmentName());
+            StreamPartitionMsgOffset completedSegmentEndOffset =
+                _streamPartitionMsgOffsetFactory.create(segmentZKMetadata.getEndOffset());
+            if (completedSegmentEndOffset.compareTo(largestStreamOffset) < 0) {
```

Comment:
Is this the only way we know that a partition group is closed? Is there a corner case that if the partitionGroup does not produce any new events after we completed the last segment, but we did not create a consuming segment (which we should have) due to some reason, then we will never flag it? Since you are taking the trouble of calling the stream, is there an API that returns ONLY active partitionGroups ?(maybe with their offsets)

Reply:
I understand. I was more intrigued that there is no way to return only live partitions from the underlying stream. @npawar is this the case?

Reply:
[Capturing our offline discussion]
You  brought up a good point that we may be catching up from a closed partition. The logic you have here seems fine to me.

Reply:
I was only concerned about `.compareTo` working for Kinesis or not but after looking at the code it should.